### PR TITLE
CSHARP-4659: Improve BsonClassMapSerializer to support stuctures

### PR DESCRIPTION
**Pull request by sanych-sun**

### CODE REVIEW

File: src/MongoDB.Bson/Serialization/BsonMemberMap.cs [hunk 1]
```diff
@@ -577,6 +577,12 @@ private static object GetDefaultValue(Type type)
                 throw new BsonSerializationException(message);
             }
 
+            if (fieldInfo.DeclaringType.IsValueType)
+            {
+                // Fallback to reflection for value type to avoid boxing-unboxing problems
```
Comment (sanych-sun) [inline src/MongoDB.Bson/Serialization/BsonMemberMap.cs:582]: I was not able to find a strict documentation for Expression.Unwrap to prof it works or find a scenarios when it not working. So I've decided to fallback to reflection, that works with once boxed struct, without unboxing it.

File: tests/MongoDB.Driver.Legacy.Tests/Jira/CSharp218Tests.cs [hunk 2]
```diff
@@ -54,18 +54,10 @@ public void TestDeserializeClassWithStructPropertyWithoutConstructorFails()
...
-                Assert.True(false, "Expected an exception to be thrown.");
-            }
-            catch (Exception ex)
-            {
-                var expectedMessage = "An error occurred while deserializing the P field of class MongoDB.Driver.Tests.Jira.CSharp218.CSharp218Tests+C: Value class MongoDB.Driver.Tests.Jira.CSharp218.CSharp218Tests+P cannot be deserialized without a constructor.";
-                Assert.IsType<FormatException>(ex);
-                Assert.IsType<BsonSerializationException>(ex.InnerException);
-                Assert.Equal(expectedMessage, ex.Message);
-            }
+
```
Comment (rstam) [inline tests/MongoDB.Driver.Legacy.Tests/Jira/CSharp218Tests.cs:57]: Remove "Fails" from the test name.

File: tests/MongoDB.Driver.Legacy.Tests/Jira/CSharp218Tests.cs [hunk 3]
```diff
@@ -74,7 +66,9 @@ public void TestDeserializeStructFails()
             _collection.RemoveAll();
             var s = new S { Id = ObjectId.GenerateNewId(), P = new P { X = 1, Y = 2 } };
             _collection.Insert(s);
-            Assert.Throws<BsonSerializationException>(() => _collection.FindOneAs<S>());
+            var result = _collection.FindOneAs<S>();
```
Comment (rstam) [inline tests/MongoDB.Driver.Legacy.Tests/Jira/CSharp218Tests.cs:69]: Remove "Fails" from the test name.

File: tests/MongoDB.Driver.Legacy.Tests/Jira/CSharp218Tests.cs (hunk 2 above)
Comment (rstam) [reply tests/MongoDB.Driver.Legacy.Tests/Jira/CSharp218Tests.cs:57]: Could also use `Remove and Sort Usings" to clean up using statements.

**Review by rstam — CHANGES_REQUESTED** [review]
Looks great. Just a few minor suggestions.

File: tests/MongoDB.Driver.Legacy.Tests/Jira/CSharp218Tests.cs (hunk 2 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Legacy.Tests/Jira/CSharp218Tests.cs:57]: Done

File: tests/MongoDB.Driver.Legacy.Tests/Jira/CSharp218Tests.cs (hunk 3 above)
Comment (sanych-sun) [reply tests/MongoDB.Driver.Legacy.Tests/Jira/CSharp218Tests.cs:69]: Done

**Review by rstam — APPROVED** [review]
LGTM

**Pull request merged**